### PR TITLE
Add container names to process minor label if possible.

### LINKED ISSUE #331: Show container name instead of process name in "Applications"
It's really hard to tell difference when you have multiple node/ruby/java/etc applications running. Could it be possible to label these with container name (if process is inside a container)?

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #331

### CODE REVIEW

**Comment by peterbourgon** [comment]
Can you provide some context here? When I'm in the applications view, I generally don't want to know about containers. Did someone ask for this feature?

File: app/api_topology_test.go [hunk 1]
```diff
@@ -67,13 +67,13 @@ func TestAPITopologyApplications(t *testing.T) {
 	defer ts.Close()
 	is404(t, ts, "/api/topology/applications/foobar")
 	{
-		body := getRawJSON(t, ts, "/api/topology/applications")
+		body := getRawJSON(t, ts, "/api/topology/containers")
```
Comment (peterbourgon) [inline app/api_topology_test.go]: Why is TestAPITopologyApplications making a GET on /api/topology/containers? Can you move this to TestAPITopologyContainers?

File: app/router.go [hunk 2]
```diff
@@ -104,7 +104,7 @@ var topologyRegistry = map[string]topologyView{
 	"applications": {
 		human:    "Applications",
 		parent:   "",
-		renderer: render.FilterUnconnected{Renderer: render.ProcessRenderer},
+		renderer: render.FilterUnconnected{Renderer: render.ProcessWithContainerNameRenderer{}},
```
Comment (peterbourgon) [inline app/router.go]: Is there a reason we wouldn't just have this be the default behavior of ProcessRenderer?

File: render/topologies.go [hunk 3]
```diff
@@ -25,6 +29,45 @@ var ProcessRenderer = MakeReduce(
...
 
+// ProcessWithContainerNameRenderer is a Renderer which produces a process
+// graph enriched with container names where appropriate
+type ProcessWithContainerNameRenderer struct{}
+
+// Render produces a process graph where the minor labels contain the
+// container name, if found.
+func (r ProcessWithContainerNameRenderer) Render(rpt report.Report) RenderableNodes {
+	var processes = ProcessRenderer.Render(rpt)
+	var containers = LeafMap{
```
Comment (peterbourgon) [inline render/topologies.go]: [Always prefer type-inferred assignment with :=](https://docs.google.com/document/d/18UTcmy8VoecEaYNb0DrI8NZrI1uJbq-XYDEko0l5Ny4/edit) unless you're doing a var block.

File: render/topologies.go [hunk 4]
```diff
@@ -25,6 +29,45 @@ var ProcessRenderer = MakeReduce(
...
+		}
+		containerID, ok := p.NodeMetadata.Metadata[docker.ContainerID]
+		if !ok {
+			continue
+		}
+		container, ok := containers[containerID]
+		if !ok {
+			continue
+		}
+		p.LabelMinor = fmt.Sprintf("%s (%s:%s)", report.ExtractHostID(p.NodeMetadata), container.LabelMajor, pid)
```
Comment (peterbourgon) [inline render/topologies.go]: Just to be clear, this is overwriting the existing minor label, which is of the format "hostname (pid)" — right?

File: app/router.go (hunk 2 above)
Comment (tomwilkie) [inline app/router.go]: The process renderer is used by the container render; there very little point in going to all the effort of labelling up the processes if we're just going to blow it away in the container renderer.

File: render/topologies.go (hunk 4 above)
Comment (tomwilkie) [inline render/topologies.go]: Yes

**Comment by tomwilkie** [comment]
@peterbourgon thanks for the feedback; PTAL

File: app/router.go (hunk 2 above)
Comment (peterbourgon) [inline app/router.go]: Ah, right, gotcha.

**Comment by peterbourgon** [comment]
:+1:

**Pull request merged**